drivers/usb/udc/udc_dwc2.c

```diff
@@ -995,22 +1008,53 @@ static inline void dwc2_handle_out_xfercompl(const struct device *dev,
 
 	evt.type = DWC2_DRV_EVT_DOUT;
 	evt.ep = ep_cfg->addr;
-	/* Assume udc buffer and endpoint config is the same as it was when
-	 * transfer was scheduled in dwc2_prep_rx(). The original transfer size
-	 * value is necessary here because controller decreases the value for
-	 * every byte stored.
+
+	/* The original transfer size value is necessary here because controller
+	 * decreases the value for every byte stored.
 	 */
-	evt.bcnt = dwc2_rx_xfer_size(priv, ep_cfg, buf) -
-		usb_dwc2_get_deptsizn_xfersize(doeptsiz);
+	bcnt = usb_dwc2_get_doeptsizn_xfersize(priv->rx_siz[ep_idx]) -
+	       usb_dwc2_get_doeptsizn_xfersize(doeptsiz);
+
+	if (is_iso) {
+		uint32_t pkts;
+		bool valid;
+
+		pkts = usb_dwc2_get_doeptsizn_pktcnt(priv->rx_siz[ep_idx]) -
+			usb_dwc2_get_doeptsizn_pktcnt(doeptsiz);
+		switch (usb_dwc2_get_doeptsizn_rxdpid(doeptsiz)) {
+		case USB_DWC2_DOEPTSIZN_RXDPID_DATA0:
+			valid = (pkts == 1);
+			break;
+		case USB_DWC2_DOEPTSIZN_RXDPID_DATA1:
+			valid = (pkts == 2);
+			break;
+		case USB_DWC2_DOEPTSIZN_RXDPID_DATA2:
+			valid = (pkts == 3);
+			break;
+		case USB_DWC2_DOEPTSIZN_RXDPID_MDATA:
+		default:
+			valid = false;
+			break;
+		}
 
-	if (priv->bufferdma) {
-		sys_cache_data_invd_range(buf->data, evt.bcnt);
+		if (!valid) {
+			if (!priv->bufferdma) {
+				/* RxFlvl added data to net buf, rollback */
+				net_buf_remove_mem(buf, bcnt);
+			}
+			/* Data is not valid, discard it */
```

Comment:
Should there be some print here?

Reply:
Definitely not, because it can then flood at 125 us rate (at High-Speed). Isochronous endpoints guarantee the data rate, but the cost is that there is no retry in case of errors.

Reply:
LOG_WRN_ONCE?

Reply:
No, I don't consider a broken isochronous transfer to be something to warn about. The deal with isochronous transfers is that application has to be prepared to deal with missing data.

Reply:
Sounds good to me.